diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7c700ab07..d8ce20ff7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -31,17 +31,6 @@ set(CMAKE_CXX_STANDARD 11)
 # plugins.
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-# Embed the installation prefix as an RPATH in the executable such that the
-# linker can find our libraries (such as libcvc4parser) when executing the cvc4
-# binary. This is for example useful when installing CVC4 with a custom prefix
-# on macOS (e.g. when using homebrew in a non-standard directory). If we do not
-# set this option, then the linker will not be able to find the required
-# libraries when trying to run CVC4.
-#
-# More information on RPATH in CMake:
-# https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/RPATH-handling
-set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
-
 #-----------------------------------------------------------------------------#
 
 include(Helpers)
@@ -390,7 +379,7 @@ endif()
 
 if(USE_CLN)
   set(GPL_LIBS "${GPL_LIBS} cln")
-  find_package(CLN 1.2.2 REQUIRED)
+  find_package(CLN 1.3.6 REQUIRED)
   set(CVC4_USE_CLN_IMP 1)
   set(CVC4_USE_GMP_IMP 0)
 else()
diff --git a/cmake/FindCLN.cmake b/cmake/FindCLN.cmake
index 7b2e6f0f4..8ff7a84ec 100644
--- a/cmake/FindCLN.cmake
+++ b/cmake/FindCLN.cmake
@@ -3,9 +3,18 @@
 # CLN_INCLUDE_DIR - the CLN include directory
 # CLN_LIBRARIES - Libraries needed to use CLN
 
-find_path(CLN_INCLUDE_DIR NAMES cln/cln.h)
-find_library(CLN_LIBRARIES NAMES cln)
+if(NOT CLN_HOME)
+  set(CLN_HOME ${PROJECT_SOURCE_DIR}/cln-1.3.6)
+endif()
 
+find_path(CLN_INCLUDE_DIR
+          NAMES cln/cln.h
+          PATHS ${CLN_HOME}/include
+          NO_DEFAULT_PATH)
+find_library(CLN_LIBRARIES
+             NAMES cln
+             PATHS ${CLN_HOME}/lib
+             NO_DEFAULT_PATH)
 
 if(CLN_INCLUDE_DIR)
   file(STRINGS
diff --git a/cmake/FindGMP.cmake b/cmake/FindGMP.cmake
index 8125a583b..9f955273d 100644
--- a/cmake/FindGMP.cmake
+++ b/cmake/FindGMP.cmake
@@ -9,7 +9,6 @@
 # fail if GMP was not found in the specified directory.
 if(NOT GMP_HOME)
   set(GMP_HOME ${PROJECT_SOURCE_DIR}/gmp-6.1.2)
-  set(CHECK_SYSTEM_VERSION TRUE)
 endif()
 
 find_path(GMP_INCLUDE_DIR
@@ -21,11 +20,6 @@ find_library(GMP_LIBRARIES
              PATHS ${GMP_HOME}/lib
              NO_DEFAULT_PATH)
 
-if(CHECK_SYSTEM_VERSION)
-  find_path(GMP_INCLUDE_DIR NAMES gmp.h gmpxx.h)
-  find_library(GMP_LIBRARIES NAMES gmp)
-endif()
-
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(GMP DEFAULT_MSG GMP_INCLUDE_DIR GMP_LIBRARIES)
 
diff --git a/contrib/get-abc b/contrib/get-abc
index fa90f240f..c34f95ce0 100755
--- a/contrib/get-abc
+++ b/contrib/get-abc
@@ -24,6 +24,11 @@ grep -v 'libSupport\.c' src/base/main/module.make.orig > src/base/main/module.ma
 cp src/base/main/mainInit.c src/base/main/mainInit.c.orig
 sed 's,\( *\)\(.*Libs_Init(\),\1//\2,;s,\( *\)\(.*Libs_End(\),\1//\2,' src/base/main/mainInit.c.orig > src/base/main/mainInit.c
 
+sed -i -z 's/#if defined(WIN32) && !defined(__cplusplus)/#if 0/3' src/base/cmd/cmd.c
+sed -i -z 's/#if defined(WIN32) && !defined(__cplusplus)/#if 0/1' src/base/cmd/cmdLoad.c
+sed -i '1s/^/define nl\n\n\nendef/' Makefile
+sed -i 's/$(VERBOSE)rm -rvf $(PROG) lib$(PROG)\.a $(OBJ) $(GARBAGE) $(OBJ:\.o=\.d)/$(VERBOSE)$(foreach file,$(PROG) lib$(PROG)\.a $(OBJ) $(GARBAGE),rm -rvf $(file)${nl})/' Makefile
+
 # Build optimized, without readline, without pthreads.
 # These aren't necessary for our usage and we don't want the dependencies.
 make -j$(nproc) libabc.a OPTFLAGS=-O ABC_USE_NO_READLINE=1 ABC_USE_NO_PTHREADS=1
diff --git a/contrib/get-antlr-3.4 b/contrib/get-antlr-3.4
index 670fc3b41..c990ed9b7 100755
--- a/contrib/get-antlr-3.4
+++ b/contrib/get-antlr-3.4
@@ -48,6 +48,8 @@ cd libantlr3c-3.4
 # warnings about the ANTLR library having moved
 PREFIX=$(realpath `pwd`/../..)
 
+sed -i '/-mingw32\*/ i\          | -msys \\' config.sub
+
 # Make antlr3debughandlers.c empty to avoid unreferenced symbols
 rm -rf src/antlr3debughandlers.c && touch src/antlr3debughandlers.c
 if [ ${MACHINE_TYPE} == 'x86_64' ]; then
@@ -73,6 +75,8 @@ mv lib/libantlr3c.a lib/libantlr3c-static.a
 cd src/libantlr3c-3.4
 make clean
 
+sed -i '/-mingw32\*/ i\          | -msys \\' config.sub
+
 if [ ${MACHINE_TYPE} == 'x86_64' ]; then
   # 64-bit stuff here
   ./configure --enable-64bit --with-pic --disable-antlrdebug --prefix="${PREFIX}" $ANTLR_CONFIGURE_ARGS $BUILD_TYPE
@@ -80,7 +84,6 @@ else
   # 32-bit stuff here
   ./configure --with-pic --disable-antlrdebug --prefix="${PREFIX}" $ANTLR_CONFIGURE_ARGS $BUILD_TYPE
 fi
-
 cp Makefile Makefile.orig
 sed 's,^\(CFLAGS = .*\),\1 -fexceptions,' Makefile.orig > Makefile
 make CFLAGS="${MAKE_CFLAGS}" CXXFLAGS="${MAKE_CXXFLAGS}" LDFLAGS="${MAKE_LDFLAGS}"
diff --git a/contrib/get-cadical b/contrib/get-cadical
index 359afaaa7..98e30da93 100755
--- a/contrib/get-cadical
+++ b/contrib/get-cadical
@@ -13,7 +13,7 @@ git clone https://github.com/arminbiere/cadical cadical
 cd cadical
 git checkout $commit
 
-CXXFLAGS="-fPIC" ./configure && make -j$(nproc)
+CXXFLAGS="-fPIC" ./configure -q && make -j$(nproc)
 
 echo
 echo "Using CaDiCaL commit $commit"
diff --git a/contrib/get-cln b/contrib/get-cln
new file mode 100755
index 000000000..1ae9948f9
--- /dev/null
+++ b/contrib/get-cln
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+source "$(dirname "$0")/get-script-header.sh"
+
+[ -z "${BUILD_TYPE}" ] && BUILD_TYPE="--disable-shared --enable-static --with-pic"
+[ -n "$HOST" ] && HOST="--host=$HOST"
+[ -z "$CLNVERSION" ] && CLNVERSION=1.3.6
+
+echo =============================================================================
+echo
+echo "Setting up CLN $CLNVERSION..."
+echo
+( set -ex
+  gmpprefix="$(pwd)/gmp-1.2"
+  mkdir cln-$CLNVERSION
+  cd cln-$CLNVERSION
+  clnprefix=`pwd` &&
+  mkdir src &&
+  cd src &&
+  webget https://www.ginac.de/CLN/cln-$CLNVERSION.tar.bz2 cln-$CLNVERSION.tar.bz2 &&
+  tar xfj cln-$CLNVERSION.tar.bz2 &&
+  cd cln-$CLNVERSION &&
+  ./configure ${HOST} --prefix="$clnprefix" --with-gmp="${gmpprefix}" ${BUILD_TYPE} &&
+  make CFLAGS="${MAKE_CFLAGS}" CXXFLAGS="${MAKE_CXXFLAGS}" LDFLAGS="${MAKE_LDFLAGS}" &&
+  make install
+) || exit 1
+
+echo
+echo ===================== Now configure CVC4 with =====================
+echo ./configure.sh --cln
diff --git a/contrib/get-drat2er b/contrib/get-drat2er
index e465ab3d4..af0abade3 100755
--- a/contrib/get-drat2er
+++ b/contrib/get-drat2er
@@ -12,6 +12,17 @@ cd drat2er
 
 git checkout api
 
+# check for broken experimental/optional on OSX
+if g++ -stc=gnu++14 -x c++ -o test-working-experimental-optional - <<'EOF'
+#include <experimental/optional>
+int main() { std::experimental::optional<int> o; }
+EOF
+then
+  rm test-working-experimental-optional
+else
+  echo "add_compile_definitions(CLI11_EXPERIMENTAL_OPTIONAL=0)" >> CMakeLists.txt
+fi
+
 mkdir build
 
 cd build
diff --git a/contrib/get-glpk-cut-log b/contrib/get-glpk-cut-log
index dcd5aac00..0d8d3721b 100755
--- a/contrib/get-glpk-cut-log
+++ b/contrib/get-glpk-cut-log
@@ -15,7 +15,7 @@ webget https://github.com/timothy-king/glpk-cut-log/archive/$commit.zip glpk-cut
 unzip glpk-cut-log-$commit.zip
 cd glpk-cut-log-$commit
 
-libtoolize
+if command -v libtoolize; then libtoolize; else glibtoolize; fi
 aclocal
 autoheader
 autoconf
diff --git a/contrib/get-gmp b/contrib/get-gmp
index 02602e456..b2b7a9b41 100755
--- a/contrib/get-gmp
+++ b/contrib/get-gmp
@@ -11,7 +11,7 @@
 
 source "$(dirname "$0")/get-script-header.sh"
 
-[ -z "${BUILD_TYPE}" ] && BUILD_TYPE="--disable-shared --enable-static"
+[ -z "${BUILD_TYPE}" ] && BUILD_TYPE="--disable-shared --enable-static --enable-fat --with-pic"
 [ -n "$HOST" ] && HOST="--host=$HOST"
 [ -z "$GMPVERSION" ] && GMPVERSION=6.1.2
 
@@ -28,6 +28,7 @@ echo
   webget https://gmplib.org/download/gmp/gmp-$GMPVERSION.tar.bz2 gmp-$GMPVERSION.tar.bz2 &&
   tar xfj gmp-$GMPVERSION.tar.bz2 &&
   cd gmp-$GMPVERSION &&
+  sed -i 's/\*-\*-mingw\*/*-*-mingw* | *-*-msys/' configure
   ./configure ${HOST} --prefix="$gmpprefix" --enable-cxx ${BUILD_TYPE} &&
   make CFLAGS="${MAKE_CFLAGS}" CXXFLAGS="${MAKE_CXXFLAGS}" LDFLAGS="${MAKE_LDFLAGS}" &&
   make install
diff --git a/src/bindings/java/CMakeLists.txt b/src/bindings/java/CMakeLists.txt
index 3423a3e1b..0acae2147 100644
--- a/src/bindings/java/CMakeLists.txt
+++ b/src/bindings/java/CMakeLists.txt
@@ -14,6 +14,7 @@ else()
   swig_add_library(cvc4jni LANGUAGE Java SOURCES ${CVC4_SWIG_INTERFACE})
 endif()
 swig_link_libraries(cvc4jni cvc4 cvc4parser ${JNI_LIBRARIRES})
+install(TARGETS cvc4jni DESTINATION lib)
 
 # Create CVC4.jar from all generated *.java files.
 set(gen_java_files
@@ -174,10 +175,7 @@ set(gen_java_files
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_ListenerCollection__Registration.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_MaybeT_CVC4__Rational_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_Type.java
-  ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_mpq_class.java
-  ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_mpz_class.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__istream.java
-  ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__mapT_CVC4__Expr_CVC4__Expr_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__ostream.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__shared_ptrT_CVC4__SygusPrintCallback_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__string.java
@@ -241,5 +239,22 @@ set(gen_java_files
   ${CMAKE_CURRENT_BINARY_DIR}/vectorVectorExpr.java
 )
 
+if(CVC4_USE_CLN_IMP)
+  list(APPEND gen_java_files
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_cln__cl_I.java
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_cln__cl_RA.java
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__mapT_CVC4__Expr_CVC4__Expr_std__lessT_CVC4__Expr_t_t.java
+  )
+elseif(CVC4_USE_GMP_IMP)
+  list(APPEND gen_java_files
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_mpq_class.java
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_mpz_class.java
+    ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__mapT_CVC4__Expr_CVC4__Expr_t.java
+  )
+endif()
+
+install(FILES ${gen_java_files} DESTINATION src/cvc4/java)
 add_jar(cvc4jar SOURCES ${gen_java_files} OUTPUT_NAME CVC4)
+install_jar(cvc4jar DESTINATION share/cvc4)
 add_dependencies(cvc4jar cvc4jni)
+
